protocol/l1-upgrades.md

```diff
@@ -0,0 +1,305 @@
+  contracts within one block.
+- **Declarative design:** As much as possible, a developer should be able to simply indicate what
+  bytecode should be deployed, and what storage values (if any) should be initialized or modified.
+
+## Non-goals
+
+This document is not concerned with:
+
+1. upgrading predeploys or other L2 contracts.
+1. modifications to the ownership system, ie, it does not attempt to modify Safes, modules or
+   guards.
+
+More over, it does not present a plan for managing contracts which are not proxied, and not
+Multichain Prepared (MCP-L1) contracts (ie. FaultDisputeGame, AnchorStateRegistry). This
+work will either include, or else depend on, all L1 contracts becoming either:
+
+1. Proxied and MCP ready
+2. Superchain-shared contracts (proxied or not)
+
+# Summary
+
+<!-- Most (if not all) documents should have a summary.
+While the length will likely be proportional to the length of the full document,
+the summary should be as succinct as possible. -->
+
+The proposed solution will perform upgrades by transferring upgrade authorization of all L1
+contracts to the release specific OPCM. Upon completion of the upgrade, this authorization will
+be returned to the Security Council.
+
+Rather than the curent `StorageSetter` based approach which simply deletes the `_initialized` slot,
+allowing for `initialize()` to be called, we take this two-step upgrade further, with a short-lived
+`Populator` contract being temporarily set as the implementation in order to set or modify any
+values as required. This has two significant benefits. First it removes the need to keep a one-time
+`initialize()` function in the run-time logic. Secondly it enables a clean separation of concerns
+between contracts which are newly deployed (requiring all values to be set), and contracts being
+upgraded (requiring only new values to be set).
+
+# Problem Statement + Context
+
+Our upgrades have the following issues:
+
+1. They are divorced from the implementation, meaning the work of writing the upgrade script occurs
+   in `superchain-ops` apart from the work of writing the updated code in the `optimism` monorepo.
+2. The act of writing the upgrade script is onerous, requiring:
+   1. A bespoke script with custom assertions.
+   2. A bespoke Validations.md document
+3. The work required to perform the upgrade is difficult for signers, as they must work through the
+   Validations.md document. This work has taken up to an hour in the past.
+4. The effort described above will scale linearly with the number of OP Chains.
+
+# Proposed Solution
+
+Each new release will have its own OP Contracts Manager which can perform the following distinct
+actions:
+
+1. Deploy new proxies for a new OP Chain (this work is already covered by [link to Blaine's doc]())
+1. Upgrade all existing OP Chains.
+1. Deploy any newly added superchain-shared contracts.
+1. Upgrade existing superchain-shared contracts.
```

Comment:
What is the absolute prestate a function of? I'd like to better understand why the above requirements are the case

> having the absolute prestate as an immutable in FaultDisputeGame seems like a pretty perfect fit for what we need, other than that it doesn't fit the typical MCP approach.

Could it be a regular storage variable with no setter method? That way we reuse the FaultDisputeGame implementation, and just put a proxy in front to have different storage values per chain

Reply:
I imagine it relates to the use of CWIA to deploy the game instances.

Reply:
The absolute prestate is basically the hash of `op-program` (converted to a cannon state snapshot). It effectively embeds the consensus rules (ie current hard fork implementation) and the chain configuration.  It can embed the chain configuration for multiple chains in the same version, but chain config changes happen at a whole bunch of times, not just when op-mainnet activates a hard fork (new chains, chains activating forks at different times etc).

We can't really use a proxy per chain because the `DisputeGameFactory` needs to be able to create new proxies when it creates a new game. I don't think the use of CWIA vs some other proxy really matters here, it's just that it's not a one-off thing, we create new instances all the time. Theoretically the `DisputeGameFactory` could store the absolute prestate for each game type but that's way outside what `DisputeGameFactory` should know about and bleeds the specific details of `FaultDisputeGame` into the otherwise pretty generic `DisputeGameFactory` (which really should have a more generic name since the "games" it creates could just as easily be for ZK proofs and wouldn't have an absolute prestate at all).

My best idea for solving this is to introduce a `ChainConfigRegistry`: https://github.com/ethereum-optimism/design-docs/pull/161 but there's details still to work out there and it hasn't been scheduled so is unlikely to be done prior to Isthmus.

We do have a good superchain-ops for the `setImplementation` calls required to upgrade fault proofs (https://github.com/ethereum-optimism/superchain-ops/pull/363) but still need something to deploy the contracts (and because of the potential for chain configs to change, that may need to happen outside of hard forks or releases of op-contracts) but it would only differ by the absolute prestate.

Reply:
For the purpose of this document, I'd summarize this as: 
1. significant work is required to make the dispute games MCP compatible, which we should not block this on. 
2. Therefore this work needs to either: 
 1. consider dispute game contracts as out of scope and needing to be upgraded via other means until they are MCP compatible.
 2. identify a workaround.
 
 
I prefer 2, using a blueprint based approach.